crates/build/src/workspace/manifest.rs

```diff
@@ -341,6 +344,57 @@ impl Manifest {
         Ok(self)
     }
 
+    /// Merge the workspace dependencies with the crate dependencies.
+    pub fn with_fixed_workspace_dependencies(
+        &mut self,
+        crate_metadata: &CrateMetadata,
+    ) -> Result<&mut Self> {
+        let workspace_manifest_path =
+            crate_metadata.cargo_meta.workspace_root.join("Cargo.toml");
+
+        // If the workspace manifest is the same as the crate manifest, there's not
+        // workspace to fix
+        if workspace_manifest_path == self.path.path {
+            return Ok(self)
```

Comment:
So if we are able to detect if there is a virtual manifest, and we are confident that this works and is well tested, I see no reason not to make this the default behaviour instead of an unstable feature.